File: servers/ktor/src/test/kotlin/com/quietmetrix/server/funnels/ActorKeyTest.kt
package com.quietmetrix.server.funnels

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * The funnel matcher needs one identity per event row: prefer the analytics-salt install
 * hash (survives across sessions), fall back to the session id for pre-migration events
 * that have no hash yet (see A2), and give up only when neither is present.
 */
class ActorKeyTest {

    @Test
    fun `prefers the install hash when present`() {
        assertEquals("abc123", actorKey(installHash = "abc123", sessionId = "sess-1"))
    }

    @Test
    fun `falls back to a prefixed session id when install hash is absent`() {
        assertEquals("sid:sess-1", actorKey(installHash = null, sessionId = "sess-1"))
    }

    @Test
    fun `falls back to session id when install hash is blank`() {
        assertEquals("sid:sess-1", actorKey(installHash = "", sessionId = "sess-1"))
    }

    @Test
    fun `is null when both are absent`() {
        assertNull(actorKey(installHash = null, sessionId = null))
    }

    @Test
    fun `is null when both are blank`() {
        assertNull(actorKey(installHash = "", sessionId = ""))
    }

    @Test
    fun `install hash never collides with a session-id fallback`() {
        // A raw session id could coincidentally equal another actor's install hash; the
        // "sid:" prefix on the fallback keeps the two identity spaces disjoint.
        val bySession = actorKey(installHash = null, sessionId = "abc123")
        val byHash = actorKey(installHash = "abc123", sessionId = "sess-1")
        assertEquals("sid:abc123", bySession)
        assertEquals("abc123", byHash)
        assert(bySession != byHash)
    }
}

File: php-hosting/tests/actorKeyTest.php
<?php

/**
 * Unit tests for actorKey() — the unit of conversion for funnel matching.
 * Mirrors servers/ktor/.../funnels/ActorKey.kt; keep the two in sync.
 *
 * Run: php php-hosting/tests/actorKeyTest.php   (exit 0 = pass, 1 = fail)
 */

require_once __DIR__ . '/../src/funnelActor.php';

$failures = 0;
function check(string $label, $expected, $actual): void {
    global $failures;
    if ($expected === $actual) {
        echo "  ok: $label\n";
        return;
    }
    $failures++;
    echo "FAIL: $label\n";
    echo '    expected: ' . var_export($expected, true) . "\n";
    echo '    actual:   ' . var_export($actual, true) . "\n";
}

check('prefers the install hash when present', 'abc123',
    actorKey('abc123', 'sess-1'));

check('falls back to a prefixed session id when install hash is absent', 'sid:sess-1',
    actorKey(null, 'sess-1'));

check('falls back to session id when install hash is blank', 'sid:sess-1',
    actorKey('', 'sess-1'));

check('is null when both are absent', null,
    actorKey(null, null));

check('is null when both are blank', null,
    actorKey('', ''));

check('install hash never collides with a session-id fallback (by session)', 'sid:abc123',
    actorKey(null, 'abc123'));

check('install hash never collides with a session-id fallback (by hash)', 'abc123',
    actorKey('abc123', 'sess-1'));

echo "\n";
if ($failures > 0) {
    echo "$failures assertion(s) failed\n";
    exit(1);
}
echo "all actorKey assertions passed\n";
